File: src/days/day3/a.py
import math
import os
import re

dir_path = os.path.dirname(os.path.realpath(__file__))
file_path = os.path.join(dir_path, "input/real.txt")
with open(file_path) as f:
    seq = f.read()

def solve(seq):
    matches = re.finditer(r"mul\(([0-9]{1,3}),([0-9]{1,3})\)", seq)
    s = [math.prod(map(int, m.groups())) for m in matches]

    return sum(s)

print(solve(seq))

File: src/days/day3/a.rs
use regex::Regex;
use std::error::Error;

use crate::utils;

pub fn solve() -> Result<(), Box<dyn Error>> {
    let program = utils::read_from_file_to_string("day3/input/real.txt".to_string())?;

    let re = Regex::new(r"mul\(([0-9]{1,3}),([0-9]{1,3})\)").unwrap();

    let matches: Result<Vec<_>, Box<dyn Error>> = re
        .captures_iter(&program)
        .map(|caps| {
            // println!("{}", &m[0]);
            let x = caps[1].parse::<i32>()?;
            let y = caps[2].parse::<i32>()?;

            Ok(x * y)
        })
        .collect();

    // println!("{:?}", matches);

    let res: i32 = matches?.into_iter().sum();

    println!("{}", res);

    Ok(())
}
